### PR TITLE
feat: allow symlink

### LINKED ISSUE #460: `ko apply --recursive -f` to follow symlinked directories
This will enable sharing *.yaml files between different deployment configurations, or with other tools (like a helm chart directory).

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

Fixes #460 

Is that what you thought? @imjasonh

### CODE REVIEW

File: pkg/commands/options/filestuff.go [hunk 1]
```diff
@@ -87,11 +87,20 @@ func EnumerateFiles(fo *FilenameOptions) chan string {
...
 						return filepath.SkipDir
 					}
 					if watcher != nil {
-						watcher.Add(path)
+						if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
+							symlink, err := filepath.EvalSymlinks(path)
+							if err != nil {
+								return err
+							}
+							watcher.Add(symlink)
```
Comment (imjasonh) [inline pkg/commands/options/filestuff.go:100]: This seems to only affect watch mode (`-W`), which is deprecated and will eventually be removed (#412).

Does this also happen to work for non-watch-mode `ko apply -f`?

Comment (developer-guy) [reply pkg/commands/options/filestuff.go]: I did something to fix that, hope it'd work 🤦🏻‍♂️

File: pkg/commands/options/filestuff.go [hunk 2]
```diff
@@ -87,11 +87,17 @@ func EnumerateFiles(fo *FilenameOptions) chan string {
 				// the directory, and we're in watch mode, then we set up a watch on the
 				// directory.
 				if fi.IsDir() {
+					symlink, err := resolveSymlink(path, fi)
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply pkg/commands/options/filestuff.go]: I've tested all of the samples and everything went well 🙋🏻‍♂️

would you mind reviewing one more time ✌️

File: pkg/commands/options/filestuff.go [hunk 3]
```diff
@@ -142,3 +155,10 @@ func EnumerateFiles(fo *FilenameOptions) chan string {
 	}()
 	return files
 }
+
+func ResolveSymlink(path string, fi os.FileInfo) (symlink string, err error) {
+	if fi.Mode()&os.ModeSymlink == os.ModeSymlink {
+		symlink, err = os.Readlink(path) // TODO: Handle error here
```
Comment (imjasonh) [inline pkg/commands/options/filestuff.go:161]: This TODO is done now right?

**Comment by imjasonh** [comment]
I think the suggestion in https://github.com/google/ko/pull/571#discussion_r792132067 is still useful, we could add a symlink in the repo and exercise this new path in an e2e test

File: pkg/commands/options/filestuff.go (hunk 3 above)
Comment (developer-guy) [reply pkg/commands/options/filestuff.go]: yes, right.

File: pkg/commands/options/filestuff.go [hunk 4]
```diff
@@ -86,17 +86,30 @@ func EnumerateFiles(fo *FilenameOptions) chan string {
...
+					return err
+				}
+
 				if fi.IsDir() {
 					if path != paths && !fo.Recursive {
 						return filepath.SkipDir
 					}
 					if watcher != nil {
-						watcher.Add(path)
+						watcher.Add(symlink)
```
Comment (imjasonh) [inline pkg/commands/options/filestuff.go:99]: Let's just remove any watcher interaction, we're going to delete it all soon anyway in #585

File: pkg/commands/options/filestuff.go [hunk 5]
```diff
@@ -86,17 +86,30 @@ func EnumerateFiles(fo *FilenameOptions) chan string {
...
+					} else {
+						if symlink != "" {
+							files <- symlink
+						}
```
Comment (imjasonh) [inline pkg/commands/options/filestuff.go:103]:
```suggestion
					} else if symlink != "" {
						files <- symlink
					}
```

File: pkg/commands/options/filestuff.go [hunk 6]
```diff
@@ -142,3 +155,10 @@ func EnumerateFiles(fo *FilenameOptions) chan string {
 	}()
 	return files
 }
+
+func ResolveSymlink(path string, fi os.FileInfo) (symlink string, err error) {
```
Comment (imjasonh) [inline pkg/commands/options/filestuff.go:159]: Can this be replaced with https://pkg.go.dev/path/filepath#EvalSymlinks ?

File: test/symlink.yaml [hunk 7]
```diff
@@ -0,0 +1 @@
+test/test.yaml
```
Comment (imjasonh) [inline test/symlink.yaml:1]: Can you add a test that checks that this symlink is followed if you `ko apply -f test/symlink.yaml` ?

File: pkg/commands/options/filestuff.go (hunk 5 above)
Comment (developer-guy) [reply pkg/commands/options/filestuff.go:103]: thank you 🙋🏻‍♂️

File: pkg/commands/options/filestuff.go (hunk 4 above)
Comment (developer-guy) [reply pkg/commands/options/filestuff.go:99]: OFC, thank you 🙋🏻‍♂️

File: pkg/commands/options/filestuff.go (hunk 6 above)
Comment (developer-guy) [reply pkg/commands/options/filestuff.go:159]: it didn't work when I switched to EvalSymlink, I don't know why 🤷🏻‍♂️

File: test/symlink.yaml (hunk 7 above)
Comment (developer-guy) [reply test/symlink.yaml]: I tried something to test this but couldn't make it work but anyways I wanted to share the code with you to see what I missed or did wrong 🙋🏻‍♂️ maybe you can help me to fix this 🙏

Comment (developer-guy) [reply test/symlink.yaml]: thanks a ton, @dentrax, he helped me to fix this. ❤️

**Comment by Dentrax** [comment]
NIT: One small concern here is that we do not support the case in which symlink points itself: `too many levels of symbolic links`

A workaround: https://github.com/fkautz/gitbom-go/pull/7/files (It would be overengineered to implement that in this project I think)

File: pkg/commands/options/filestuff.go (hunk 6 above)
Comment (imjasonh) [reply pkg/commands/options/filestuff.go:154]: I'd like to understand better why that's the case. This is getting fairly complicated, and I'd rather depend on stdlib methods as much as possible. If we still need to write significant code to handle symlinks, I'm a bit more inclined not to support symlinks for now.

**Comment by developer-guy** [comment]
kindly ping 🙋🏻‍♂️

**Pull request closed without merging**